### ISSUE TITLE
`defaults()` transforms null list of string into an empty list of string

### ISSUE THREAD

**Issue by mderriey-prospa** [hashicorp/terraform#29306]
### Terraform Version

```
Terraform v1.0.3
on windows_amd64
```

### Terraform Configuration Files

`main.tf`

```terraform
terraform {
  required_version = ">= 1.0.3"
  experiments      = [module_variable_optional_attrs]
}

variable "input" {
  type = list(object({
    name            = string
    optional_string = optional(string)
    optional_list   = optional(list(string))
  }))
}

locals {
  defaults_input = defaults(var.input, {
    "optional_string" = "Default value"
  })
}

output "input" {
  value = var.input
}

output "defaults_input" {
  value = local.defaults_input
}
```

`terraform.tfvars`

```terraform
input = [
  {
    "name" = "My name"
  }
]
```

### Debug Output

<https://gist.github.com/mderriey-prospa/0ff18ed7babbe6363ace0ff114dfa5bd>.

### Crash Output

N/A.

### Expected Behavior

The `default_list` property should stay `null` after being processed by the `defaults` function.

### Actual Behavior

The `default_list` property is an empty list after being processed by the `defaults` function.

### Steps to Reproduce

1. Create the Terraform configuration files mentioned above in a directory.
2. Run `terraform apply -auto-approve`.

### Additional Context

N/A.

### References

N/A.

**Comment by mderriey-prospa** [comment]
I should not that in this case, the workaround I found is have another local to override the empty lists, e.g.

```terraform
locals {
  // Run it through defaults which will transform null into empty lists
  intermediate_input = defaults(var.input, {
    "optional_string" = "Default value"
  })

  // Run the result through a merge statement to replace empty lists with null
  input = [
    for input in local.intermediate_input :
    merge(input, {
      "optional_list": length(input.optional_list) == 0 ? null : input.optional_list
    })
  ]
}
```

**Comment by korinne** [comment]
Hi, we recently released the [Terraform v1.3 alpha](https://releases.hashicorp.com/terraform/1.3.0-alpha20220608), which includes the ability to mark object type attributes as optional and set default values.

I'm following up with issues that have provided feedback on the previous experiment, and wanted to invite you all to try the alpha and provide any feedback on this updated design. You can learn more/add comments[ here](https://discuss.hashicorp.com/t/request-for-feedback-optional-object-type-attributes-with-defaults-in-v1-3-alpha/40550). Thank you so much in advance, and hope you like the feature!

**Comment by apparentlymart** [comment]
Hello again!

As a result of this and other feedback that the `defaults` function wasn't able to treat collection values in a way that was intuitive for all situations, we removed the function from the final incarnation of the feature and instead integrated support for default values directly into the type constraint syntax.

The example in the original issue comment could be adapted to the final design of the feature as follows:

```hcl
variable "input" {
  type = list(object({
    name            = string
    optional_string = optional(string, "Default value")
    optional_list   = optional(list(string))
  }))
}
```

Any optional attribute which doesn't have a specified default value will default to `null` as before, so with the input value included in the original writeup the final value of `var.input` in this case would be:

```hcl
tolist([
  {
    name            = "My name"
    optional_string = "Default value"
    optional_list   = null /* of list(string) type */
  },
])
```

If an author instead _wanted_ `optional_list` to default to an empty list, they could declare it as `optional(list(string), [])`.

This functionality is already available in the release candidate of Terraform v1.3.0 and will be included in the forthcoming final release. Thanks!
